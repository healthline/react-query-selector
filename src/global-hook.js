export function findFiber(hostNode) {
  return Object.values(window.__REACT_DEVTOOLS_GLOBAL_HOOK__._renderers).reduce(
    (prev, renderer) => prev || renderer.findFiberByHostInstance(hostNode),
    undefined
  );
}

export function fiberRoots() {
  return [].concat(
    ...Object.values(window.__REACT_DEVTOOLS_GLOBAL_HOOK__._fiberRoots).map(
      root => Array.from(root)
    )
  );
}

export let ReactTypeOfWork;

function setupReactDom({ version }) {
  // **********************************************************
  // The section below is copy-pasted from files in React DevTools repo.
  // Keep it in sync, and add version guards if it changes.
  // https://github.com/facebook/react-devtools/blob/master/backend/attachRendererFiber.js
  // **********************************************************
  // The section below is copy-pasted from files in React repo.
  // Keep it in sync, and add version guards if it changes.
  // **********************************************************
  if (version >= '16.6.0') {
    ReactTypeOfWork = {
      FunctionalComponent: 0,
      ClassComponent: 1,
      IndeterminateComponent: 2,
      HostRoot: 3,
      HostPortal: 4,
      HostComponent: 5,
      HostText: 6,
      Fragment: 7,
      Mode: 8,
      ContextConsumer: 9,
      ContextProvider: 10,
      ForwardRef: 11,
      Profiler: 12,
      SuspenseComponent: 13,
      MemoComponent: 14,
      SimpleMemoComponent: 15,
      LazyComponent: 16,
      IncompleteClassComponent: 17,
      CoroutineComponent: -1, // Removed
      CoroutineHandlerPhase: -1, // Removed
      YieldComponent: -1 // Removed
    };
  } else if (version >= '16.4.3') {
    ReactTypeOfWork = {
      FunctionalComponent: 0,
      FunctionalComponentLazy: 1,
      ClassComponent: 2,
      ClassComponentLazy: 3,
      IndeterminateComponent: 4,
      HostRoot: 5,
      HostPortal: 6,
      HostComponent: 7,
      HostText: 8,
      Fragment: 9,
      Mode: 10,
      ContextConsumer: 11,
      ContextProvider: 12,
      ForwardRef: 13,
      ForwardRefLazy: 14,
      Profiler: 15,
      PlaceholderComponent: 16
    };
  } else {
    ReactTypeOfWork = {
      IndeterminateComponent: 0,
      FunctionalComponent: 1,
      FunctionalComponentLazy: -1, // Doesn't exist yet
      ClassComponent: 2,
      ClassComponentLazy: -1, // Doesn't exist yet
      HostRoot: 3,
      HostPortal: 4,
      HostComponent: 5,
      HostText: 6,
      CoroutineComponent: 7,
      CoroutineHandlerPhase: 8,
      YieldComponent: 9,
      Fragment: 10,
      Mode: 11,
      ContextConsumer: 12,
      ContextProvider: 13,
      ForwardRef: 14,
      ForwardRefLazy: -1, // Doesn't exist yet
      Profiler: 15,
      Placeholder: 16
    };
  }
}

if (typeof __REACT_DEVTOOLS_GLOBAL_HOOK__ !== 'undefined') {
  Object.values(__REACT_DEVTOOLS_GLOBAL_HOOK__._renderers).forEach(
    setupReactDom
  );

  __REACT_DEVTOOLS_GLOBAL_HOOK__.on('renderer', ({ renderer }) =>
    setupReactDom(renderer)
  );
} else {
  // Provide our own hook implementation if not running in a devtools environment
  let hook;
  window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = hook = {
    _renderers: [],
    _fiberRoots: [],
    inject: instance => {
      hook._renderers.push(instance);
      hook._fiberRoots.push([]);
      setupReactDom(instance);
      return hook._renderers.length - 1;
    },
    onCommitFiberRoot: (rendererId, root) => {
      const mountedRoots = hook._fiberRoots[rendererId];
      const current = root.current;
      const isKnownRoot = mountedRoots.includes(root);
      const isUnmounting =
        current.memoizedState == null || current.memoizedState.element == null;
      // Keep track of mounted roots so we can hydrate when DevTools connect.
      if (!isKnownRoot && !isUnmounting) {
        hook._fiberRoots[rendererId].push(root);
      } else if (isKnownRoot && isUnmounting) {
        hook.onCommitFiberUnmount(rendererId, root);
      }
    },
    onCommitFiberUnmount: (rendererId, root) => {
      hook._fiberRoots[rendererId] = hook._fiberRoots[rendererId].filter(
        current => current !== root
      );
    },
    supportsFiber: true
  };
}
