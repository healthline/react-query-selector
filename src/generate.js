import { findFiber } from './global-hook';
import { getTypeName } from './adapter';
import { generateFullDomMap } from './map';

export function generatePath(hostNode) {
  // Use the React DOM to generate selector, if possible.
  let path;
  let component = findFiber(hostNode) || hostNode._reactInternalFiber;
  if (component) {
    const domMap = generateFullDomMap();

    let typeName = getTypeName(component);
    let mapped = domMap[typeName].find(
      mapItem => mapItem.node.stateNode === component.stateNode
    );

    path = mapped.path
      .reverse()
      .filter(
        ({ type, selector, ofType }, index) =>
          !index || /^</.test(selector) || ofType
      )
      .map(({ type, selector }) => [selector, domMap[type].length]);

    let singleElement = path.findIndex(
      ([name, count]) => /^</.test(name) && count === 1
    );
    if (singleElement >= 0) {
      path = path.slice(0, singleElement + 1);
    }

    return path
      .map(([name]) => name)
      .reverse()
      .join(' ');
  }
}
