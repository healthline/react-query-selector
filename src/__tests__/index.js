import React from 'react';
import ReactQuerSelector, { querySelector, querySelectorAll } from '..';
import { generatePath } from '../generate';
import MyComponent from '../../fixtures/my-component';
import { setupReactDom } from '../global-hook';

const App = () => {
  return (
    <div>
      <p>React here!</p>
    </div>
  );
};

let hook;
global.__REACT_DEVTOOLS_GLOBAL_HOOK__ = hook = {
  _renderers: [],
  _fiberRoots: [],
  inject: instance => {
    hook._renderers.push(instance);
    hook._fiberRoots.push([]);
    setupReactDom(instance);
    return hook._renderers.length - 1;
  },
  onCommitFiberRoot: jest.fn((rendererId, root) => {
    hook._fiberRoots[rendererId].push(root);
  }),
  onCommitFiberUnmount: (rendererId, root) => {
    hook._fiberRoots[rendererId] = hook._fiberRoots[rendererId].filter(
      current => current !== root
    );
  },
  supportsFiber: true
};

describe('react-query-selector', () => {
  const container = document.body.appendChild(document.createElement('div'));
  beforeEach(() => {
    hook.onCommitFiberRoot.mockClear();

    ReactDOM.unmountComponentAtNode(container);
    container.innerHTML = '';
  });

  // Must come after global above to ensure init state.
  // We may want to move the mock above into a setup file ast some point
  const ReactDOM = require('react-dom');

  test('it should inject into global hook', () => {
    ReactDOM.render(<App />, container);

    expect(hook.onCommitFiberRoot).toHaveBeenCalledTimes(1);
    expect(hook._renderers).toHaveLength(1);
  });

  describe('generatePath', () => {
    it('should provide path for named elements', () => {
      ReactDOM.render(<App />, container);

      const el = container.querySelector('p');
      const path = generatePath(el);
      expect(path).toMatchSnapshot();
      expect(querySelector(path)).toBe(el);
    });
    it('should provide path for anonymous elements', () => {
      ReactDOM.render(<p>value</p>, container);

      const el = container.querySelector('p');
      const path = generatePath(el);
      expect(path).toMatchSnapshot();
      expect(querySelector(path)).toBe(el);
    });

    it('should provide path for nested elements', () => {
      ReactDOM.render(<MyComponent />, container);

      const el = container.querySelector('p');
      const path = generatePath(el);
      expect(path).toMatchSnapshot();
      expect(querySelector(path)).toBe(el);

      const path2 = generatePath(el.previousElementSibling);
      expect(path2).toMatchSnapshot();
      expect(querySelector(path2)).toBe(el.previousElementSibling);
    });

    it('should provide unique path for multiple elements', () => {
      ReactDOM.render(
        <span>
          <MyComponent />
          <MyComponent />
        </span>,
        container
      );

      const el = container.querySelector('p');
      const path = generatePath(el);
      expect(path).toMatchSnapshot();
      expect(querySelector(path)).toBe(el);

      const path2 = generatePath(el.previousElementSibling);
      expect(path2).toMatchSnapshot();
      expect(querySelector(path2)).toBe(el.previousElementSibling);
    });
  });

  describe('querySelectorAll', () => {
    it('should return multiple elements', () => {
      ReactDOM.render(
        <span>
          <MyComponent />
          <MyComponent />
        </span>,
        container
      );

      expect(querySelectorAll('<Functional>')).toHaveLength(2);
    });
  });

  describe('scoping', () => {
    it('should return multiple elements', () => {
      ReactDOM.render(
        <span>
          <MyComponent />
          <MyComponent />
        </span>,
        container
      );

      expect(querySelectorAll('<Functional>', document)).toHaveLength(2);
      expect(
        querySelectorAll('<Functional>', document.createElement('div'))
      ).toHaveLength(0);
      expect(
        querySelectorAll('<Functional>', querySelector('<MyComponent>'))
      ).toHaveLength(1);
      expect(
        querySelectorAll('<Functional>', querySelector('span'))
      ).toHaveLength(2);
    });
  });
});
