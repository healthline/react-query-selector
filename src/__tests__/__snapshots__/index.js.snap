// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-query-selector generatePath should provide path for anonymous elements 1`] = `"p"`;

exports[`react-query-selector generatePath should provide path for named elements 1`] = `"<App> p"`;

exports[`react-query-selector generatePath should provide path for nested elements 1`] = `"<Functional> p"`;

exports[`react-query-selector generatePath should provide path for nested elements 2`] = `"<Functional> div:nth-of-type(2)"`;

exports[`react-query-selector generatePath should provide unique path for multiple elements 1`] = `"<MyComponent>:nth-of-type(1) <Functional> p"`;

exports[`react-query-selector generatePath should provide unique path for multiple elements 2`] = `"<MyComponent>:nth-of-type(1) <Functional> div:nth-of-type(2)"`;
